File: base_common/src/main/java/com/lazyxu/base/test/ActivityStack.java
//
// Source code recreated from a .class file by IntelliJ IDEA
// (powered by Fernflower decompiler)
//

package com.lazyxu.base.test;

import android.app.Activity;
import android.content.Context;
import android.content.Intent;
import android.os.Process;
import java.util.Stack;

public class ActivityStack {
  private Stack<Activity> mActivities = new Stack();
  private static final String TAG = "ActivityStack";
  private static ActivityStack INSTANCE;

  private ActivityStack() {
  }

  public static ActivityStack getInstance() {
    if (INSTANCE == null) {
      INSTANCE = new ActivityStack();
    }

    return INSTANCE;
  }
  public void popAllActivity(boolean isForceClose) {
    while(this.mActivities.size() > 0) {
      this.popActivity();
    }
    if (isForceClose) {
      Process.killProcess(Process.myPid());
      System.exit(-1);
    }
  }

  public void popActivity() {
    if (this.mActivities != null && !this.mActivities.isEmpty()) {
      Activity activity = this.mActivities.pop();
      if (activity != null) {
        activity.finish();
        activity = null;
      }
    }
  }
  public void popAllActivityExceptTop() {
    while(this.mActivities.size() > 1) {
      Activity activity = this.mActivities.get(0);
      this.mActivities.remove(activity);
      if (activity != null) {
        activity.finish();
        activity = null;
      }
    }
  }






  public void popActivity(Activity activity) {
    if (this.mActivities != null && !this.mActivities.isEmpty() && this.mActivities.contains(activity)) {
      this.mActivities.remove(activity);
      if (activity != null) {
        activity.finish();
        activity = null;
      }
    }

  }

  public Activity topActivity() {
    return this.mActivities.empty() ? null : (Activity)this.mActivities.lastElement();
  }

  public void pushActivity(Activity activity) {
    this.mActivities.add(activity);
  }




  public void reStartApp(Context mContext, Class<?> className) {
    Intent intent = new Intent(mContext, className);
    intent.addFlags(335544320);
    mContext.startActivity(intent);
    Process.killProcess(Process.myPid());
  }

  public void popActivityUntilCls(Class<?> clz) {
    while(this.mActivities.size() > 1 && this.topActivity().getClass() != clz) {
      this.popActivity();
    }

  }

  public void popAllActivityExceptBottom() {
    int size = this.mActivities.size();

    for(int i = size - 1; i >= 1; --i) {
      Activity activity = (Activity)this.mActivities.get(i);
      this.mActivities.remove(activity);
      if (activity != null) {
        activity.finish();
      }
    }

  }

  public int size() {
    return this.mActivities.size();
  }

  public Activity activityAt(int position) {
    return position < this.mActivities.size() ? (Activity)this.mActivities.elementAt(position) : null;
  }
}

File: base_common/src/main/java/com/lazyxu/base/utils/ActivityStack.kt
package com.lazyxu.base.utils

import android.app.Activity
import android.content.Context
import android.os.Process
import android.view.inputmethod.InputMethodManager
import java.util.*
import kotlin.system.exitProcess

/**
 * User:Lazy_xu
 * Data:2019/10/10
 * Description: kotlin的静态内部类方式的单例模式
 * FIXME
 */
class ActivityStack private constructor() {
    //stack栈，先进后出,继承Vector线程安全
    private var mActivities = Stack<Activity>()

    companion object {
        val INSTANCE: ActivityStack by lazy(mode = LazyThreadSafetyMode.SYNCHRONIZED) { ActivityStack() }
    }

    fun addActivity(activity: Activity) {
        this.mActivities.add(activity)
    }

    fun removeActivity(activity: Activity) {
        with(activity) {
            if (!mActivities.isEmpty() && mActivities.contains(activity)) {
                mActivities.remove(this)
                this.finish()
            }
        }
    }

    fun popAllActivity(isForceClose: Boolean) {
        while (this.mActivities.size > 0) {
            this.popActivity()
        }
        if (isForceClose) {
            Process.killProcess(Process.myPid())
            exitProcess(-1)
        }
    }

    private fun popActivity() {
        if (!this.mActivities.isEmpty()) {
            //pop栈顶元素出栈
            (mActivities.pop() as Activity).finish()
        }
    }

    fun popAllActivityExceptTop() {
        while (this.mActivities.size > 1) {
            val activity: Activity = this.mActivities[0] as Activity
            this.mActivities.remove(activity)
            activity.finish()
        }
    }


    private fun hideSoftKeyBoard(activity: Activity) {
        val localView = activity.currentFocus
        val imm = activity.getSystemService(Context.INPUT_METHOD_SERVICE) as InputMethodManager
        if (localView != null && imm != null) {
            imm.hideSoftInputFromWindow(localView.windowToken, InputMethodManager.HIDE_NOT_ALWAYS)
        }
    }
}
